### PR TITLE
[numpy] numpy einsum

### PR DESCRIPTION
**Pull request by hzfan**
## Description ##
numpy compatible [einsum](https://docs.scipy.org/doc/numpy/reference/generated/numpy.einsum.html)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- add einsum
- add einsum doc
- add path optimization for imperative mode
- add benchmark
- add forward and backward test

## Comments ##
- support both explicit and implicit mode
- support multiple operands
- support broadcast
- support 0-dim NDArray and 0-size NDArray

## TODO ##
- [x] translate einsum_path into c, so that optimization can work in symbol mode
- [x] take advantage of tensordot when it is merged
- [x] support "greedy" optimization
- [ ] support "optimal" optimization

Thank @reminisce and @haojin2 for guidance and review.

### CODE REVIEW

**Comment by hzfan** [comment]
@mxnet-label-bot add [Numpy]

File: src/operator/numpy/np_einsum_op-inl.h [hunk 1]
```diff
@@ -0,0 +1,772 @@
...
+#include "../elemwise_op_common.h"
+
+
+namespace mxnet {
+namespace op {
+
+#define NPY_MAXDIMS 32
+#define NPY_MAXARGS 32
+
+
```
Comment (haojin2) [inline src/operator/numpy/np_einsum_op-inl.h]: one less blank line here.

File: src/operator/numpy/np_einsum_op-inl.h [hunk 2]
```diff
@@ -0,0 +1,772 @@
...
+  int ndim = shape.ndim();
+  CHECK_GE(odim, ndim);
+  TShape ret(odim, 1);
+  for (int idim = 0; idim < ndim; ++idim) {
+    ret[idim] = shape[idim];
+  }
+  return ret;
+}
+
+
```
Comment (haojin2) [inline src/operator/numpy/np_einsum_op-inl.h]: one less blank line here

File: src/operator/numpy/np_einsum_op.cc [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+                                          output_labels);
+    CHECK_GE(ndim_output, 0);
+  }
+
+  // std::cout << "output_labels" << std::endl;
+  // for (int i = 0; i < ndim_output; ++i) {
+  //   std::cout << output_labels[i] << " ";
+  // }
+  // std::cout << std::endl;
+  // std::cout << "max_broadcast = " << max_broadcast << std::endl;
```
Comment (haojin2) [inline src/operator/numpy/np_einsum_op.cc]: remove dead code

File: python/mxnet/numpy/multiarray.py [hunk 4]
```diff
@@ -2873,3 +2873,225 @@ def arcsin(x, out=None, **kwargs):
...
+    array([[10., 28., 46., 64.],
+           [13., 40., 67., 94.]])
+
+    Chained array operations. For more complicated contractions, speed ups
+    might be achieved by repeatedly computing a 'greedy' path or pre-computing the
+    'optimal' path and repeatedly applying it, using an
+    `einsum_path` insertion (since version 1.12.0). Performance improvements can be
+    particularly significant with larger arrays:
+
+    >>> a = np.ones(64).reshape(2,4,8)
```
[the remark this answers is not in the document]
Comment (hzfan) [reply python/mxnet/numpy/multiarray.py]: I haven't benchmarked yet. The benchmark here is from official np.einsum. I will remove them.

File: python/mxnet/numpy/multiarray.py [hunk 5]
```diff
@@ -2873,3 +2873,225 @@ def arcsin(x, out=None, **kwargs):
...
+    array([[10., 28., 46., 64.],
+           [13., 40., 67., 94.]])
+    >>> np.einsum('k...,jk', a, b)
+    array([[10., 28., 46., 64.],
+           [13., 40., 67., 94.]])
+
+    Chained array operations. For more complicated contractions, speed ups
+    might be achieved by repeatedly computing a 'greedy' path or pre-computing the
+    'optimal' path and repeatedly applying it, using an
+    `einsum_path` insertion (since version 1.12.0). Performance improvements can be
```
[the remark this answers is not in the document]
Comment (hzfan) [reply python/mxnet/numpy/multiarray.py]: removed

File: src/operator/numpy/np_einsum_op-inl.h [hunk 6]
```diff
@@ -0,0 +1,772 @@
...
+                                   char *labels,
+                                   TShape* newshape,
+                                   TShape* newstride) {
+  using namespace mxnet_op;
+  int idim, ndim, icombine, combineoffset;
+  int icombinemap[NPY_MAXDIMS];
+  int newdim;
+
+  const TShape& shape = op.shape_;
+  TShape stride;
```
[the remark this answers is not in the document]
Comment (hzfan) [reply src/operator/numpy/np_einsum_op-inl.h]: done

File: src/operator/numpy/np_einsum_op-inl.h [hunk 7]
```diff
@@ -0,0 +1,772 @@
...
+  return ndim;
+}
+
+
+inline void get_combined_dims_view(const TBlob& op, int iop,
+                                   char *labels,
+                                   TShape* newshape,
+                                   TShape* newstride) {
+  using namespace mxnet_op;
+  int idim, ndim, icombine, combineoffset;
```
[the remark this answers is not in the document]
Comment (hzfan) [reply src/operator/numpy/np_einsum_op-inl.h]: done

File: src/operator/numpy/np_einsum_op.cc [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+ */
+
+#include "./np_einsum_op-inl.h"
+#include <stdlib.h>
+#include <string.h>
+
+namespace mxnet {
+namespace op {
+
+inline bool NumpyEinsumShape(const nnvm::NodeAttrs& attrs,
```
[the remark this answers is not in the document]
Comment (hzfan) [reply src/operator/numpy/np_einsum_op.cc]: removed

File: src/operator/numpy/np_einsum_op-inl.h (hunk 2 above)
Comment (hzfan) [reply src/operator/numpy/np_einsum_op-inl.h]: done.

File: src/operator/numpy/np_einsum_op-inl.h (hunk 1 above)
Comment (hzfan) [reply src/operator/numpy/np_einsum_op-inl.h]: done

File: src/operator/numpy/np_einsum_op.cc (hunk 3 above)
Comment (hzfan) [reply src/operator/numpy/np_einsum_op.cc]: done

**Comment by hzfan** [comment]
cherry-picked from  #15565. Tensordot optimization has been enabled. @reminisce

**Comment by hzfan** [comment]
PRed to the master branch. Closing this one.

**Pull request closed without merging**